plenoptic/synthesize/geodesic.py

```diff
@@ -1,155 +1,340 @@
+        optimizer: {'Adam', torch.optim.Optimizer}, optional
+            choice of algorithm that will perform the search
+            if an optimizer is passed, its `params` argument should be set to
+            `self.x`, where self refers to a previously initialized Geodesic
+            class.
+        lmbda: float, optional
+            strength of the regularizer that enforces the image range in [0, 1]
+            (strictly positive by default)
+        tol: float, optional
+            tolerance threshold used to terminate algorithm before `max_iter`
+            if the optimization stopped making progress
+        seed: int, optional
+            set the random number generator
+        verbose: bool, optional
+            storing information along the run of the optimization algorithm
+        """
+        self.lmbda = lmbda
+        self.verbose = verbose
+        if tol is None:
+            tol = self.pixelfade.norm() / 1e4 * (1 + 5 ** .5) / 2
+        print(f"\n threshold for delta_x, tolerance = {tol:.5e}")
 
         torch.manual_seed(seed)
-        self.optimizer = optim.Adam([self.x], lr=learning_rate, amsgrad=True)
-
-        self.loss = []
-        self.t1 = time.time()
-
-        for i in range(max_iter):
-            loss = self._optimizer_step(i, max_iter)
-            self.loss.append(loss.item())
-
-        self.gamma.copy_(self.img.data.squeeze()).add_(0.5).mul_(255)
-
-        return self.gamma
+        if optimizer == 'Adam':
+            self.optimizer = optim.Adam([self.x],
+                                        lr=learning_rate, amsgrad=True)
+        elif isinstance(optimizer, optim.Optimizer):
+            self.optimizer = optimizer
+
+        delta_x = torch.ones(1)
+        i = 0
+        with tqdm(range(max_iter)) as pbar:
+            # project onto set of representational geodesics
+            while delta_x > tol and i < max_iter:
+                delta_x = self._optimizer_step(pbar)
+                i += 1
+        self.populate_geodesic()
+
+    def plot_loss(self):
+        fig, ax = plt.subplots(1, 1)
+        ax.plot(self.loss)
+        ax.set(yscale='log',
+               xlabel='iter step',
+               ylabel='loss value')
+        return fig
+
+    def populate_geodesic(self):
+        """Help format the current optimization variable `x` into a geodesic
+        attribute that can be used later.
+
+        It joins the endpoints, reshapes the variable into a sequence
+        of images, detaches the gradients and clips the range to [0, 1].
+        """
+        self.geodesic = torch.cat([self.xA, self.x, self.xB]
+                                  ).reshape(
+                                (self.n_steps+1, *self.image_shape[1:])
+                                            ).clamp(0, 1).detach()
+
+    def plot_deviation_from_line(self, video=None, figsize=(7, 5)):
+        """visual diagnostic of geodesic linearity in representation space.
```

Comment:
Since this is a diagnostic plot, can we get an explanation for what it should look like, if things are working? (straight line, right?)